File: typescript/tests/unit/mobile.test.ts
/**
 * Unit tests for Mobile module.
 * Following TDD principles - tests first, then implementation.
 */

import { Mobile } from '../../src/mobile/mobile';

// Mock Session interface
interface MockSession {
  callMcpTool: jest.Mock;
  getAPIKey: () => string;
  sessionId: string;
}

describe('Mobile', () => {
  let mockSession: MockSession;
  let mobile: Mobile;

  beforeEach(() => {
    mockSession = {
      callMcpTool: jest.fn(),
      getAPIKey: () => 'test-api-key',
      sessionId: 'test-session-id'
    };
    mobile = new Mobile(mockSession as any);
  });

  describe('Touch Operations', () => {
    test('tap should call MCP tool with correct parameters', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        errorMessage: ''
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.tap(100, 200);

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('tap', {
        x: 100,
        y: 200
      });
      expect(result.success).toBe(true);
      expect(result.requestId).toBe('test-123');
    });

    test('swipe should call MCP tool with correct parameters', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.swipe(100, 100, 200, 200, 500);

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('swipe', {
        start_x: 100,
        start_y: 100,
        end_x: 200,
        end_y: 200,
        duration_ms: 500
      });
      expect(result.success).toBe(true);
    });

    test('swipe should use default duration when not provided', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.swipe(100, 100, 200, 200);

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('swipe', {
        start_x: 100,
        start_y: 100,
        end_x: 200,
        end_y: 200,
        duration_ms: 300
      });
    });
  });

  describe('Input Operations', () => {
    test('inputText should call MCP tool with correct parameters', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.inputText('Hello Mobile');

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('input_text', {
        text: 'Hello Mobile'
      });
      expect(result.success).toBe(true);
    });

    test('sendKey should call MCP tool with correct parameters', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.sendKey(4); // BACK key

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('send_key', {
        key: 4
      });
      expect(result.success).toBe(true);
    });
  });

  describe('UI Element Operations', () => {
    test('getClickableUIElements should call MCP tool and return elements', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        data: JSON.stringify([
          {
            text: 'Button 1',
            className: 'android.widget.Button',
            bounds: { left: 0, top: 0, right: 100, bottom: 50 }
          }
        ])
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.getClickableUIElements(5000);

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('get_clickable_ui_elements', {
        timeout_ms: 5000
      });
      expect(result.elements).toHaveLength(1);
      expect(result.elements[0].text).toBe('Button 1');
    });

    test('getAllUIElements should call MCP tool and return all elements', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        data: JSON.stringify([
          {
            text: 'Text View',
            className: 'android.widget.TextView',
            bounds: { left: 0, top: 0, right: 200, bottom: 30 }
          }
        ])
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.getAllUIElements(3000);

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('get_all_ui_elements', {
        timeout_ms: 3000,
        format: 'json'
      });
      expect(result.format).toBe('json');
      expect(typeof result.raw).toBe('string');
      expect(result.raw).toContain('android.widget.TextView');
      expect(result.elements).toHaveLength(1);
      expect(result.elements[0].className).toBe('android.widget.TextView');
    });

    test('getAllUIElements should support XML format and return raw', async () => {
      // Arrange
      const xml = "<?xml version='1.0' encoding='UTF-8' standalone='yes' ?><hierarchy rotation=\"0\"></hierarchy>";
      const mockResult = {
        success: true,
        requestId: 'test-xml-123',
        data: xml
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.getAllUIElements(3000, 'xml');

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('get_all_ui_elements', {
        timeout_ms: 3000,
        format: 'xml'
      });
      expect(result.success).toBe(true);
      expect(result.format).toBe('xml');
      expect(result.raw.startsWith('<?xml')).toBe(true);
      expect(result.elements).toEqual([]);
    });
  });

  describe('App Management Operations', () => {
    test('getInstalledApps should call MCP tool and return apps', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        data: JSON.stringify([
          {
            name: 'Calculator',
            startCmd: 'com.android.calculator2',
            workDirectory: ''
          }
        ])
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.getInstalledApps(false, true, true);

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('get_installed_apps', {
        start_menu: false,
        desktop: true,
        ignore_system_apps: true
      });
      expect(result.apps).toHaveLength(1);
      expect(result.apps[0].name).toBe('Calculator');
    });

    test('startApp should call MCP tool with correct parameters', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        data: JSON.stringify([
          {
            pid: 1234,
            pname: 'com.android.calculator2'
          }
        ])
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.startApp('com.android.calculator2', '', '');

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('start_app', {
        start_cmd: 'com.android.calculator2',
        work_directory: '',
        activity: ''
      });
      expect(result.processes).toHaveLength(1);
      expect(result.processes[0].pid).toBe(1234);
    });

    test('stopAppByCmd should call MCP tool with correct parameters', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.stopAppByCmd('am force-stop com.android.calculator2');

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('stop_app_by_cmd', {
        stop_cmd: 'am force-stop com.android.calculator2'
      });
      expect(result.success).toBe(true);
    });
  });

  describe('Screen Operations', () => {
    test('screenshot should call MCP tool and return URL', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        data: 'https://example.com/mobile-screenshot.png'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.screenshot();

      // Assert
      expect(mockSession.callMcpTool).toHaveBeenCalledWith('system_screenshot', {});
      expect(result.data).toBe('https://example.com/mobile-screenshot.png');
    });
  });

  describe('Error Handling', () => {
    test('should handle MCP tool errors gracefully', async () => {
      // Arrange
      const mockError = new Error('MCP tool failed');
      mockSession.callMcpTool.mockRejectedValue(mockError);

      // Act
      const result = await mobile.tap(100, 200);

      // Assert
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain('MCP tool failed');
    });

    test('should handle invalid JSON response for UI elements', async () => {
      // Arrange
      const mockResult = {
        success: true,
        requestId: 'test-123',
        data: 'invalid json'
      };
      mockSession.callMcpTool.mockResolvedValue(mockResult);

      // Act
      const result = await mobile.getClickableUIElements(5000);

      // Assert
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain('Failed to parse');
    });
  });

  describe('ADB Connection', () => {
    test.skip('getAdbUrl should fail on non-mobile environment', async () => {
      // Note: This test is skipped because getAdbUrl no longer validates image ID
      // on the client side. It directly calls the getLink API which handles
      // environment validation server-side.
      //
      // Full getAdbUrl functionality is covered by integration tests.
    });

    test('getAdbUrl should succeed on mobile environment', async () => {
      // Arrange
      const expectedUrl = 'adb connect xx.xx.xx.xx:xxxxx';
      const mockGetAdbLink = jest.fn().mockResolvedValue({
        body: {
          success: true,
          requestId: 'test-adb-123',
          data: { url: expectedUrl }
        }
      });
      const mockClient = { getAdbLink: mockGetAdbLink };
      const mockAgentBay = {
        getClient: jest.fn().mockReturnValue(mockClient)
      };
      const mockSessionWithImage = {
        ...mockSession,
        sessionId: 'test-session-123',
        imageId: 'mobile_latest',
        getAPIKey: jest.fn().mockReturnValue('test-api-key'),
        getAgentBay: jest.fn().mockReturnValue(mockAgentBay)
      };
      const mobileWithImage = new Mobile(mockSessionWithImage as any);
      const adbkeyPub = 'QAAAAM0muSn7yQCY...test_key...EAAQAA=';

      // Act
      const result = await mobileWithImage.getAdbUrl(adbkeyPub);

      // Assert
      expect(mockGetAdbLink).toHaveBeenCalledTimes(1);
      const callArgs = mockGetAdbLink.mock.calls[0][0];
      expect(callArgs.authorization).toBe('Bearer test-api-key');
      expect(callArgs.sessionId).toBe('test-session-123');
      expect(callArgs.option).toContain('adbkey_pub');
      expect(result.success).toBe(true);
      expect(result.data).toBe(expectedUrl);
      expect(result.url).toBe(expectedUrl);
      expect(result.requestId).toBe('test-adb-123');
    });

    test('getAdbUrl should handle getLink errors', async () => {
      // Arrange
      const mockGetAdbLink = jest.fn().mockRejectedValue(new Error('Network error'));
      const mockClient = { getAdbLink: mockGetAdbLink };
      const mockAgentBay = { 
        getClient: jest.fn().mockReturnValue(mockClient)
      };
      const mockSessionWithImage = {
        ...mockSession,
        imageId: 'mobile_latest',
        getAPIKey: jest.fn().mockReturnValue('test-api-key'),
        getAgentBay: jest.fn().mockReturnValue(mockAgentBay)
      };
      const mobileWithImage = new Mobile(mockSessionWithImage as any);
      const adbkeyPub = 'test_key_456';

      // Act
      const result = await mobileWithImage.getAdbUrl(adbkeyPub);

      // Assert
      expect(result.success).toBe(false);
      expect(result.errorMessage).toContain('Failed to get ADB URL');
      expect(result.errorMessage).toContain('Network error');
    });
  });
}); 

File: golang/tests/pkg/unit/mobile_test.go
package agentbay_test

import (
	"testing"

	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/mobile"
	"github.com/aliyun/wuying-agentbay-sdk/golang/pkg/agentbay/models"
	"github.com/aliyun/wuying-agentbay-sdk/golang/tests/pkg/unit/mock"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/suite"
)

type MobileTestSuite struct {
	suite.Suite
	mockSession *mock.MockSessionForComputer
	mobile      *mobile.Mobile
}

func (suite *MobileTestSuite) SetupTest() {
	suite.mockSession = &mock.MockSessionForComputer{}
	suite.mobile = mobile.NewMobile(suite.mockSession)
}

func (suite *MobileTestSuite) TearDownTest() {
	suite.mockSession.AssertExpectations(suite.T())
}

// Test Tap functionality
func (suite *MobileTestSuite) TestTap_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-tap-123",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "tap", map[string]interface{}{
		"x": 100,
		"y": 200,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.Tap(100, 200)

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-tap-123", result.RequestID)
	assert.Empty(suite.T(), result.ErrorMessage)
}

func (suite *MobileTestSuite) TestTap_McpToolError() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      false,
		RequestID:    "test-tap-error",
		ErrorMessage: "Tap failed",
	}

	suite.mockSession.On("CallMcpTool", "tap", map[string]interface{}{
		"x": 100,
		"y": 200,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.Tap(100, 200)

	// Assert
	assert.False(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-tap-error", result.RequestID)
	assert.Equal(suite.T(), "Tap failed", result.ErrorMessage)
}

// Test Swipe functionality
func (suite *MobileTestSuite) TestSwipe_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-swipe-456",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "swipe", map[string]interface{}{
		"start_x":     100,
		"start_y":     200,
		"end_x":       300,
		"end_y":       400,
		"duration_ms": 500,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.Swipe(100, 200, 300, 400, 500)

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-swipe-456", result.RequestID)
}

func (suite *MobileTestSuite) TestSwipe_QuickSwipe() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-swipe-quick",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "swipe", map[string]interface{}{
		"start_x":     50,
		"start_y":     100,
		"end_x":       150,
		"end_y":       100,
		"duration_ms": 200,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.Swipe(50, 100, 150, 100, 200)

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-swipe-quick", result.RequestID)
}

// Test InputText functionality
func (suite *MobileTestSuite) TestInputText_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-input-789",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "input_text", map[string]interface{}{
		"text": "Hello Mobile",
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.InputText("Hello Mobile")

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-input-789", result.RequestID)
}

func (suite *MobileTestSuite) TestInputText_EmptyText() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-input-empty",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "input_text", map[string]interface{}{
		"text": "",
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.InputText("")

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-input-empty", result.RequestID)
}

// Test SendKey functionality
func (suite *MobileTestSuite) TestSendKey_BackKey() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-key-back",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "send_key", map[string]interface{}{
		"key": 4, // BACK key
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.SendKey(4)

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-key-back", result.RequestID)
}

func (suite *MobileTestSuite) TestSendKey_HomeKey() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-key-home",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "send_key", map[string]interface{}{
		"key": 3, // HOME key
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.SendKey(3)

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-key-home", result.RequestID)
}

// Test GetClickableUIElements functionality
func (suite *MobileTestSuite) TestGetClickableUIElements_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-elements-abc",
		Data:         `[{"text": "Button1", "bounds": {"left": 0, "top": 0, "right": 100, "bottom": 50}}, {"text": "Button2", "bounds": {"left": 0, "top": 60, "right": 100, "bottom": 110}}]`,
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "get_clickable_ui_elements", map[string]interface{}{
		"timeout_ms": 2000,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.GetClickableUIElements(2000)

	// Assert
	assert.Equal(suite.T(), "test-elements-abc", result.RequestID)
	assert.Len(suite.T(), result.Elements, 2)
	assert.Equal(suite.T(), "Button1", result.Elements[0].Text)
	assert.Equal(suite.T(), "Button2", result.Elements[1].Text)
	assert.Equal(suite.T(), 0, result.Elements[0].Bounds.Left)
	assert.Equal(suite.T(), 100, result.Elements[0].Bounds.Right)
}

func (suite *MobileTestSuite) TestGetClickableUIElements_InvalidJSON() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-elements-invalid",
		Data:         `[{"text": "Button1", "bounds": invalid}]`,
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "get_clickable_ui_elements", map[string]interface{}{
		"timeout_ms": 2000,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.GetClickableUIElements(2000)

	// Assert
	assert.Equal(suite.T(), "test-elements-invalid", result.RequestID)
	assert.Contains(suite.T(), result.ErrorMessage, "failed to parse UI elements")
}

// Test GetAllUIElements functionality
func (suite *MobileTestSuite) TestGetAllUIElements_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-all-elements-def",
		Data:         `[{"text": "Label", "className": "TextView", "bounds": {"left": 0, "top": 0, "right": 200, "bottom": 30}}]`,
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "get_all_ui_elements", map[string]interface{}{
		"timeout_ms": 3000,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.GetAllUIElements(3000)

	// Assert
	assert.Equal(suite.T(), "test-all-elements-def", result.RequestID)
	assert.Len(suite.T(), result.Elements, 1)
	assert.Equal(suite.T(), "Label", result.Elements[0].Text)
	assert.Equal(suite.T(), "TextView", result.Elements[0].ClassName)
}

// Test GetInstalledApps functionality
func (suite *MobileTestSuite) TestGetInstalledApps_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-apps-ghi",
		Data:         `[{"name": "Calculator", "start_cmd": "com.android.calculator2"}, {"name": "Camera", "start_cmd": "com.android.camera"}]`,
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "get_installed_apps", map[string]interface{}{
		"start_menu":         false,
		"desktop":            true,
		"ignore_system_apps": true,
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.GetInstalledApps(false, true, true)

	// Assert
	assert.Equal(suite.T(), "test-apps-ghi", result.RequestID)
	assert.Len(suite.T(), result.Apps, 2)
	assert.Equal(suite.T(), "Calculator", result.Apps[0].Name)
	assert.Equal(suite.T(), "com.android.calculator2", result.Apps[0].StartCmd)
	assert.Equal(suite.T(), "Camera", result.Apps[1].Name)
}

// Test StartApp functionality
func (suite *MobileTestSuite) TestStartApp_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-start-jkl",
		Data:         `[{"pname": "calculator", "pid": 1234}]`,
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "start_app", map[string]interface{}{
		"start_cmd":      "com.android.calculator2",
		"work_directory": "",
		"activity":       ".MainActivity",
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.StartApp("com.android.calculator2", "", ".MainActivity")

	// Assert
	assert.Equal(suite.T(), "test-start-jkl", result.RequestID)
	assert.Len(suite.T(), result.Processes, 1)
	assert.Equal(suite.T(), "calculator", result.Processes[0].PName)
	assert.Equal(suite.T(), 1234, result.Processes[0].PID)
}

func (suite *MobileTestSuite) TestStartApp_WithWorkDirectory() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-start-workdir",
		Data:         `[{"pname": "myapp", "pid": 5678}]`,
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "start_app", map[string]interface{}{
		"start_cmd":      "com.example.myapp",
		"work_directory": "/data/app",
		"activity":       ".SplashActivity",
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.StartApp("com.example.myapp", "/data/app", ".SplashActivity")

	// Assert
	assert.Equal(suite.T(), "test-start-workdir", result.RequestID)
	assert.Len(suite.T(), result.Processes, 1)
	assert.Equal(suite.T(), "myapp", result.Processes[0].PName)
}

// Test StopAppByPName functionality
func (suite *MobileTestSuite) TestStopAppByPName_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-stop-mno",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "stop_app_by_pname", map[string]interface{}{
		"pname": "calculator",
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.StopAppByPName("calculator")

	// Assert
	assert.True(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-stop-mno", result.RequestID)
}

func (suite *MobileTestSuite) TestStopAppByPName_AppNotFound() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      false,
		RequestID:    "test-stop-notfound",
		ErrorMessage: "App not found",
	}

	suite.mockSession.On("CallMcpTool", "stop_app_by_pname", map[string]interface{}{
		"pname": "nonexistent",
	}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.StopAppByPName("nonexistent")

	// Assert
	assert.False(suite.T(), result.Success)
	assert.Equal(suite.T(), "test-stop-notfound", result.RequestID)
	assert.Equal(suite.T(), "App not found", result.ErrorMessage)
}

// Test Screenshot functionality
func (suite *MobileTestSuite) TestScreenshot_Success() {
	// Arrange
	expectedResult := &models.McpToolResult{
		Success:      true,
		RequestID:    "test-screenshot-pqr",
		Data:         "https://example.com/mobile-screenshot.png",
		ErrorMessage: "",
	}

	suite.mockSession.On("CallMcpTool", "system_screenshot", map[string]interface{}{}).Return(expectedResult, nil)

	// Act
	result := suite.mobile.Screenshot()

	// Assert
	assert.Equal(suite.T(), "test-screenshot-pqr", result.RequestID)
	assert.Equal(suite.T(), "https://example.com/mobile-screenshot.png", result.Data)
	assert.Empty(suite.T(), result.ErrorMessage)
}

// Test comprehensive scenarios
func (suite *MobileTestSuite) TestCompleteUserFlow() {
	// This test simulates a complete user flow: tap -> input -> swipe -> screenshot

	// Step 1: Tap on input field
	suite.mockSession.On("CallMcpTool", "tap", map[string]interface{}{
		"x": 100,
		"y": 200,
	}).Return(&models.McpToolResult{
		Success:   true,
		RequestID: "flow-tap",
	}, nil)

	// Step 2: Input text
	suite.mockSession.On("CallMcpTool", "input_text", map[string]interface{}{
		"text": "Test Input",
	}).Return(&models.McpToolResult{
		Success:   true,
		RequestID: "flow-input",
	}, nil)

	// Step 3: Swipe to scroll
	suite.mockSession.On("CallMcpTool", "swipe", map[string]interface{}{
		"start_x":     100,
		"start_y":     400,
		"end_x":       100,
		"end_y":       200,
		"duration_ms": 300,
	}).Return(&models.McpToolResult{
		Success:   true,
		RequestID: "flow-swipe",
	}, nil)

	// Step 4: Take screenshot
	suite.mockSession.On("CallMcpTool", "system_screenshot", map[string]interface{}{}).Return(&models.McpToolResult{
		Success:   true,
		RequestID: "flow-screenshot",
		Data:      "https://example.com/flow-screenshot.png",
	}, nil)

	// Execute the flow
	tapResult := suite.mobile.Tap(100, 200)
	inputResult := suite.mobile.InputText("Test Input")
	swipeResult := suite.mobile.Swipe(100, 400, 100, 200, 300)
	screenshotResult := suite.mobile.Screenshot()

	// Assert all steps succeeded
	assert.True(suite.T(), tapResult.Success)
	assert.True(suite.T(), inputResult.Success)
	assert.True(suite.T(), swipeResult.Success)
	assert.Equal(suite.T(), "https://example.com/flow-screenshot.png", screenshotResult.Data)
}

// Run the test suite
func TestMobileTestSuite(t *testing.T) {
	suite.Run(t, new(MobileTestSuite))
}
